src/math.cpp

```diff
@@ -355,83 +527,144 @@ void randomseed(const CallContext& ctx) {
 }
 
 auto sin(const CallContext& ctx) -> Value {
-    return math_helper(ctx, static_cast<double (*)(double)>(&std::sin), "sin");
+    auto origin = Origin(UnaryOrigin{
+        .val = make_owning<Value>(ctx.arguments().get(0)),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& new_value,
+                      const Value& old_value) -> std::optional<SourceChangeTree> {
+            Number n = std::get<Number>(new_value);
+            return old_value.force(std::asin(n.value));
+        }});
+    return math_helper(ctx, static_cast<double (*)(double)>(&std::sin), "sin").with_origin(origin);
 }
 
 auto sqrt(const CallContext& ctx) -> Value {
-    return math_helper(ctx, static_cast<double (*)(double)>(&std::sqrt), "sqrt");
+    auto origin = Origin(UnaryOrigin{
+        .val = make_owning<Value>(ctx.arguments().get(0)),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& new_value,
+                      const Value& old_value) -> std::optional<SourceChangeTree> {
+            Number n = std::get<Number>(new_value);
+            return old_value.force(n.value * n.value);
+        }});
+    return math_helper(ctx, static_cast<double (*)(double)>(&std::sqrt), "sqrt")
+        .with_origin(origin);
 }
 
 auto tan(const CallContext& ctx) -> Value {
-    return math_helper(ctx, static_cast<double (*)(double)>(&std::tan), "tan");
+    auto origin = Origin(UnaryOrigin{
+        .val = make_owning<Value>(ctx.arguments().get(0)),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& new_value,
+                      const Value& old_value) -> std::optional<SourceChangeTree> {
+            Number n = std::get<Number>(new_value);
+            return old_value.force(std::atan(n.value));
+        }});
+    return math_helper(ctx, static_cast<double (*)(double)>(&std::tan), "tan").with_origin(origin);
 }
 
 auto to_integer(const CallContext& ctx) -> Value {
+    auto origin = Origin(UnaryOrigin{
+        .val = make_owning<Value>(ctx.arguments().get(0)),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& new_value,
+                      const Value& old_value) -> std::optional<SourceChangeTree> {
+            if (new_value == Nil()) {
+                return std::nullopt; // if result is Nil, converting to an integer was not possible.
+                                     // so it isnt possible to revert anything.
+            } else {
+                return old_value.force(new_value);
+            }
+            // Second solution (i dont know which is the correct one)
+            // please select the right one in the review and delete the wrong one
+            // return old_value.force(new_value);
```

Comment:
I think the first version is fine. Because forcing to nil would mean we would change the input to an invalid value. Which I think we should not do (and it's not obvious what that value would even be).